core/utils/attribute_helper.hpp

```diff
@@ -0,0 +1,78 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2020 ArangoDB GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Andrey Abramov
+////////////////////////////////////////////////////////////////////////////////
+
+#ifndef IRESEARCH_ATTRIBUTE_HELPER_H
+#define IRESEARCH_ATTRIBUTE_HELPER_H
+
+#include "attribute_provider.hpp"
+
+namespace iresearch {
+
+template<typename T>
+struct attribute_ptr {
+  T* ptr{};
+
+  constexpr attribute_ptr() noexcept = default;
+
+  // Intentionally implicit
+  constexpr attribute_ptr(T& v) noexcept : ptr{&v} {}
+
+  // Intentionally implicit
+  constexpr attribute_ptr(T* v) noexcept : ptr{v} {}
+
+  // Intentionally implicit
+  constexpr operator attribute_ptr<attribute>() const noexcept {
+    return attribute_ptr<attribute>{ptr};
+  }
+};
+
+template<typename T>
+struct type<attribute_ptr<T>> : type<T> {};
+
+namespace detail {
+
+template<size_t I, typename... T>
+constexpr attribute_ptr<attribute> get_mutable_helper(
+    std::tuple<T...>& t, type_info::type_id id) noexcept {
+  auto& v = std::get<I>(t);
+  if (type<std::remove_reference_t<decltype(v)>>::id() == id) {
+    return v;
```

Comment:
Why don't you return `static_cast<attribute*>(&v);` instead of `attribute_ptr<attribute>`?

Reply:
I think that's due to `std::get<I>(t)` might return either `T&` or `attribute_ptr<T>&`.